[BUG] REORG TABLE fails with NoSuchMethodError in Delta 4.0.0 with Spark 4.0.1

## Bug

#### Which Delta project/connector is this regarding?
<!--
Please add the component selected below to the beginning of the issue title
For example: [BUG][Spark] Title of my issue
-->

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Describe the problem

The REORG TABLE APPLY (PURGE) command fails with NoSuchMethodError for tables with deletion vectors enabled. This error seems to occur only in Delta 4.0.0 with Spark 4.0.1. It was not observed with Delta 4.0.0 and Spark 4.0.0. This error seems because of the signature change of `ParquetToSparkSchemaConverter` between Spark 4.0.0 and 4.0.1.

#### Steps to reproduce

Run this code.

```python
        from delta import DeltaTable
        spark.range(1000).withColumnRenamed('id', 'l_id').join(spark.range(10)).write.partitionBy('l_id').format('delta').option('delta.enableDeletionVectors', True).save('/tmp/range_dvp4')
        spark.sql("DELETE FROM delta.`/tmp/range_dvp4` WHERE id = 1")
        spark.sql("REORG table delta.`/tmp/range_dvp4` APPLY (PURGE)")
```

#### Observed results

The following error was thrown:

```
java.lang.NoSuchMethodError: 'void org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.<init>(boolean, boolean, boolean, boolean, boolean)
```

#### Expected results

The reorg command should succeed.

#### Further details

Here is a full stack trace.

```
E                   Caused by: java.lang.NoSuchMethodError: 'void org.apache.spark.sql.execution.datasources.parquet.ParquetToSparkSchemaConverter.<init>(boolean, boolean, boolean, boolean, boolean)'
E                   	at org.apache.spark.sql.delta.commands.ReorgTableHelper.filterParquetFiles(ReorgTableHelper.scala:149)
E                   	at org.apache.spark.sql.delta.commands.ReorgTableHelper.filterParquetFiles$(ReorgTableHelper.scala:122)
E                   	at org.apache.spark.sql.delta.commands.DeltaPurgeOperation.filterParquetFiles(DeltaReorgTableCommand.scala:108)
E                   	at org.apache.spark.sql.delta.commands.ReorgTableHelper.$anonfun$filterParquetFilesOnExecutors$1(ReorgTableHelper.scala:118)
E                   	at org.apache.spark.sql.execution.MapPartitionsExec.$anonfun$doExecute$3(objects.scala:198)
E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:901)
E                   	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:901)
E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
E                   	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
E                   	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:374)
E                   	at org.apache.spark.rdd.RDD.iterator(RDD.scala:338)
E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:93)
E                   	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:171)
E                   	at org.apache.spark.scheduler.Task.run(Task.scala:147)
E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$5(Executor.scala:647)
E                   	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:80)
E                   	at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:77)
E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:99)
E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:650)
E                   	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
E                   	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
E                   	... 1 more
```

### Environment information

* Delta Lake version: 4.0.0
* Spark version: 4.0.1
* Scala version: 2.13

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [x] No. I cannot contribute a bug fix at this time.

--- a/spark/src/main/scala/org/apache/spark/sql/delta/commands/ReorgTableHelper.scala
+++ b/spark/src/main/scala/org/apache/spark/sql/delta/commands/ReorgTableHelper.scala
@@ -26,6 +26,7 @@ import org.apache.hadoop.fs.{FileStatus, Path}
 
 import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.execution.datasources.parquet.{ParquetFileFormat, ParquetToSparkSchemaConverter}
+import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types.{AtomicType, StructField, StructType}
 import org.apache.spark.util.SerializableConfiguration
 
@@ -107,25 +108,27 @@ trait ReorgTableHelper extends Serializable {
       filterFileFn: StructType => Boolean): Seq[AddFile] = {
 
     val serializedConf = new SerializableConfiguration(snapshot.deltaLog.newDeltaHadoopConf())
-    val assumeBinaryIsString = spark.sessionState.conf.isParquetBinaryAsString
-    val assumeInt96IsTimestamp = spark.sessionState.conf.isParquetINT96AsTimestamp
     val dataPath = new Path(snapshot.deltaLog.dataPath.toString)
 
     import org.apache.spark.sql.delta.implicits._
 
     files.toDF(spark).as[AddFile].mapPartitions { iter =>
-        filterParquetFiles(iter.toList, dataPath, serializedConf.value, ignoreCorruptFiles,
-          assumeBinaryIsString, assumeInt96IsTimestamp)(filterFileFn).toIterator
+      val sqlConf = SparkSession.active.sessionState.conf
+      filterParquetFiles(
+        sqlConf,
+        iter.toList,
+        dataPath,
+        serializedConf.value,
+        ignoreCorruptFiles)(filterFileFn).toIterator
     }.collect()
   }
 
   protected def filterParquetFiles(
+      sqlConf: SQLConf,
       files: Seq[AddFile],
       dataPath: Path,
       configuration: Configuration,
-      ignoreCorruptFiles: Boolean,
-      assumeBinaryIsString: Boolean,
-      assumeInt96IsTimestamp: Boolean)(
+      ignoreCorruptFiles: Boolean)(
       filterFileFn: StructType => Boolean): Seq[AddFile] = {
     val nameToAddFileMap = generateCandidateFileMap(dataPath, files)
 
@@ -145,8 +148,10 @@ trait ReorgTableHelper extends Serializable {
       fileStatuses.toList,
       ignoreCorruptFiles)
 
-    val converter =
-      new ParquetToSparkSchemaConverter(assumeBinaryIsString, assumeInt96IsTimestamp)
+    // Spark 4.0.1 changed the primary ctor signature (added a param), which breaks binary
+    // compatibility for code compiled against Spark 4.0.0. Use the stable SQLConf-based ctor
+    // that takes the current SparkSession's SQLConf instead.
+    val converter = new ParquetToSparkSchemaConverter(sqlConf)
 
     val filesNeedToRewrite = footers.filter { footer =>
       val fileSchema = ParquetFileFormat.readSchemaFromFooter(footer, converter)

